### PR TITLE
JENKINS-54439 Jcasc compatibility

### PR DESCRIPTION
**Pull request by timja**
Migrates from a "Plugin" to a Descriptor
Removes custom databinding in configure

Updates jelly to use data binding
Update parent pom
Remove findbugs config, it's now in parent pom, (using spotbugs instead)
Upgrades checkstyle to allow using lamdas

### CODE REVIEW

**Comment by timja** [comment]
@rsandell mind taking a look please?

**Comment by luispollo** [comment]
@timja Thanks for contributing this! Been looking forward to this feature. @rsandell Any chance you'd get this merged any time soon?

File: src/test/java/com/sonyericsson/jenkins/plugins/bfa/PluginImplHudsonTest.java [hunk 1]
```diff
@@ -98,53 +96,6 @@ public void testGetKnowledgeBaseDescriptors() throws Exception {
...
     }
 
-    /**
-     * Tests {@link PluginImpl#configure(org.kohsuke.stapler.StaplerRequest, net.sf.json.JSONObject)}. with the same
-     * KnowledgeBase as before.
-     *
-     * @throws Exception if so.
-     */
-    @Test
-    public void testConfigure() throws Exception {
```
Comment (rsandell) [inline src/test/java/com/sonyericsson/jenkins/plugins/bfa/PluginImplHudsonTest.java]: Why are these removed?

File: src/test/java/com/sonyericsson/jenkins/plugins/bfa/PluginImplHudsonTest.java [hunk 2]
```diff
@@ -153,48 +104,21 @@ public void testConfigureLowScanThreads() throws Exception {
...
-    }
-
-    /**
-     * Tests {@link PluginImpl#configure(org.kohsuke.stapler.StaplerRequest, net.sf.json.JSONObject)}. with the same
-     * KnowledgeBase type but different configuration.
-     *
-     * @throws Exception if so.
-     */
-    @Test
-    public void testConfigureConvertSameType() throws Exception {
```
Comment (rsandell) [inline src/test/java/com/sonyericsson/jenkins/plugins/bfa/PluginImplHudsonTest.java]: Why is this test removed?

File: src/test/java/com/sonyericsson/jenkins/plugins/bfa/jcasc/ConfigurationAsCodeLocalTest.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import static org.hamcrest.core.Is.is;
+import static org.hamcrest.core.IsInstanceOf.instanceOf;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Checks configuration as code integration for local DB.
+ */
+public class ConfigurationAsCodeLocalTest {
+
+    static final String NO_CAUSES_MESSAGE = "No problems were identified. If you know why this problem "
```
Comment (rsandell) [inline src/test/java/com/sonyericsson/jenkins/plugins/bfa/jcasc/ConfigurationAsCodeLocalTest.java]: This looks like the default message, so doesn't really test that the value was correctly set :)

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/PluginImpl.java [hunk 4]
```diff
@@ -239,6 +235,76 @@ public ScanOnDemandVariables getSodVariables() {
...
+     *
+     * @return minimum sod worker threads
+     */
+    @Restricted(NoExternalUse.class)
+    public int getMinimumSodWorkerThreads() {
+        if (sodVariables == null) {
+            return ScanOnDemandVariables.DEFAULT_MINIMUM_SOD_WORKER_THREADS;
+        }
+
+        return sodVariables.getMinimumSodWorkerThreads();
```
Comment (rsandell) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/PluginImpl.java]: Why did you had to make the change in `ScanOnDemandBaseActionTest` if the potential NPE is handled here?

File: src/test/java/com/sonyericsson/jenkins/plugins/bfa/sod/ScanOnDemandBaseActionTest.java [hunk 5]
```diff
@@ -66,6 +67,11 @@
     @Rule
     public JenkinsRule j = new JenkinsRule();
 
+    @Before
+    public void setup() {
+        PluginImpl.getInstance().setSodVariables(new ScanOnDemandVariables());
```
Comment (rsandell) [inline src/test/java/com/sonyericsson/jenkins/plugins/bfa/sod/ScanOnDemandBaseActionTest.java]: Smells to me like something is wrong here if you have to add this?

File: swe_checkstyle.xml [hunk 6]
```diff
@@ -9,27 +9,26 @@
 <!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.2//EN" "http://www.puppycrawl.com/dtds/configuration_1_2.dtd">
 <module name="Checker">
     <property name="severity" value="error"/>
-    <module name="FileLength"/>
```
Comment (rsandell) [inline swe_checkstyle.xml]: There are a lot of changes in this file that I don't 
understand why.
Is it newline changes or something?

Comment (timja) [reply swe_checkstyle.xml]: could you review the checkstyle changes in this PR please?
https://github.com/jenkinsci/build-failure-analyzer-plugin/pull/98

checkstyle had to be upgraded so that java 8 code would work.

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/PluginImpl.java [hunk 7]
```diff
@@ -576,83 +673,26 @@ public KnowledgeBase getKnowledgeBase() {
...
-                logger.log(Level.SEVERE, "Could not start new knowledge base, reverting ", e);
-                save();
-                return;
-            }
+        if (existingKb != null && !existingKb.equals(knowledgeBase)) {
             if (o.getBoolean("convertOldKb")) {
                 try {
-                    base.convertFrom(knowledgeBase);
+                    knowledgeBase.start();
+                    knowledgeBase.convertFrom(existingKb);
```
[the remark this answers is not in the document]
Comment (timja) [reply src/main/java/com/sonyericsson/jenkins/plugins/bfa/PluginImpl.java]: It doesn't belong there from my understanding of what this does, seems like a hack here anyway. Should be an action from some button I would have thought...

File: src/test/java/com/sonyericsson/jenkins/plugins/bfa/PluginImplHudsonTest.java (hunk 1 above)
Comment (timja) [reply src/test/java/com/sonyericsson/jenkins/plugins/bfa/PluginImplHudsonTest.java]: I "think" it was because this test was testing form binding and form binding is now using databound semantics and tested via the jcasc test, re-testing databinding here would be awkward.

File: src/test/java/com/sonyericsson/jenkins/plugins/bfa/PluginImplHudsonTest.java (hunk 2 above)
Comment (timja) [reply src/test/java/com/sonyericsson/jenkins/plugins/bfa/PluginImplHudsonTest.java]: I "think" it was because this test was testing form binding and form binding is now using databound semantics and tested via the jcasc test, re-testing databinding here would be awkward.

File: src/test/java/com/sonyericsson/jenkins/plugins/bfa/jcasc/ConfigurationAsCodeLocalTest.java (hunk 3 above)
Comment (timja) [reply src/test/java/com/sonyericsson/jenkins/plugins/bfa/jcasc/ConfigurationAsCodeLocalTest.java]:
```suggestion
    static final String NO_CAUSES_MESSAGE = "No problems were identified. Please contribute failure causes to help others "
```

File: src/test/java/com/sonyericsson/jenkins/plugins/bfa/sod/ScanOnDemandBaseActionTest.java (hunk 5 above)
Comment (timja) [reply src/test/java/com/sonyericsson/jenkins/plugins/bfa/sod/ScanOnDemandBaseActionTest.java]: ScanOnDemandVariables aren't set by default in the current setup, probably needs a readResolve fix to initialise them if they aren't set

File: swe_checkstyle.xml (hunk 6 above)
Comment (rsandell) [reply swe_checkstyle.xml]: merged that

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/PluginImpl.java (hunk 4 above)
Comment (timja) [reply src/main/java/com/sonyericsson/jenkins/plugins/bfa/PluginImpl.java]: ScanOnDemandVariables aren't set by default in the current setup, probably needs a readResolve fix to initialise them if they aren't set. I'll take a look

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/PluginImpl.java [hunk 8]
```diff
@@ -144,9 +153,30 @@
      */
     private ScanOnDemandVariables sodVariables;
 
+    /**
+     * Default constructor.
+     */
+    @DataBoundConstructor
+    public PluginImpl() {
+        load();
```
[the remark this answers is not in the document]
Comment (timja) [reply src/main/java/com/sonyericsson/jenkins/plugins/bfa/PluginImpl.java]: I've removed the call from start, I don't think there's any real harm in it getting called extra times, but normally it gets called in a constructor for a GlobalConfiguration (IIRC)

File: src/test/java/com/sonyericsson/jenkins/plugins/bfa/sod/ScanOnDemandBaseActionTest.java (hunk 5 above)
Comment (timja) [reply src/test/java/com/sonyericsson/jenkins/plugins/bfa/sod/ScanOnDemandBaseActionTest.java]: I've pushed a fix, readResolve doesn't seem to work in the test, I added readResolve and a default initialiser to be safe

**Comment by timja** [comment]
Weirdly changing from a PluginImpl to a GlobalConfiguration seems to have some impact on classloading. Currently this branch re-introduces the deserialisation issue (albeit with a different error).

Deep in the jackson code:
```
getTypeFactory().findClass(subClass)
```
throws:
```
java.lang.ClassNotFoundException: com.sonyericsson.jenkins.plugins.bfa.model.indication.BuildLogIndication
```

**Comment by timja** [comment]
Looks like a classloader needs passing, https://jenkins.io/doc/developer/plugin-development/dependencies-and-class-loading/

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/PluginImpl.java (hunk 4 above)
Comment (timja) [reply src/main/java/com/sonyericsson/jenkins/plugins/bfa/PluginImpl.java]: The test didn't call these getters they were only called by jelly.
I've removed these getters now anyway.

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/PluginImpl.java [hunk 9]
```diff
@@ -239,6 +235,76 @@ public ScanOnDemandVariables getSodVariables() {
...
+    public int getSodWaitForJobShutdownTimeout() {
+        if (sodVariables == null) {
+            return ScanOnDemandVariables.DEFAULT_SOD_WAIT_FOR_JOBS_SHUTDOWN_TIMEOUT;
+        }
+
+        return sodVariables.getSodWaitForJobShutdownTimeout();
+    }
+
+    /**
+     * Used by jelly for loading the view, it can't seem to handle retrieving from the nested field.
```
[the remark this answers is not in the document]
Comment (timja) [reply src/main/java/com/sonyericsson/jenkins/plugins/bfa/PluginImpl.java]: Got it working thanks, example I worked from was: https://stackoverflow.com/a/52589178/4951015

**Comment by timja** [comment]
All review comments should be fixed / answered, could you please re-review, thanks.

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/PluginImpl.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (rsandell) [reply src/main/java/com/sonyericsson/jenkins/plugins/bfa/PluginImpl.java]: The idea from the beginning was that you might start with a local file database, and then later you might want some statistics or you add another Jenkins master and want to share the database. So doing the conversion on the configuration change was sensible in our opinion.
It would be nice if that was supported for JCasC as well, but can be added later I think.

[the remark this answers is not in the document]
Comment (timja) [reply src/main/java/com/sonyericsson/jenkins/plugins/bfa/PluginImpl.java]: I see, can we do it separately please

File: src/test/java/com/sonyericsson/jenkins/plugins/bfa/PluginImplHudsonTest.java (hunk 2 above)
Comment (rsandell) [reply src/test/java/com/sonyericsson/jenkins/plugins/bfa/PluginImplHudsonTest.java]: Why?
Having the plugin behave as normal when upgrading is essential. And just because the databinding works in JCasC doesn't mean it works for Jelly form binding :)
And it is definately testing in the area where there are changes, so if they doon't pass the something has regressed?

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/PluginImpl.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (rsandell) [reply src/main/java/com/sonyericsson/jenkins/plugins/bfa/PluginImpl.java]: Yes, as I said it can be done at a later date. Doesn't even have to be this year ;)

File: src/test/java/com/sonyericsson/jenkins/plugins/bfa/PluginImplHudsonTest.java (hunk 2 above)
Comment (timja) [reply src/test/java/com/sonyericsson/jenkins/plugins/bfa/PluginImplHudsonTest.java]: Restored the test, this one shouldn't have been removed

**Comment by rsandell** [comment]
I'll release all this as a beta version because I'm going on vacation today and will be gone a month. If people are happy then I can release it as a full version. But I don't think I would dare to release it as a full version now when I will not be available to review and release any bugfixes that might arise.
